src/gerbil/runtime/table.ss

```diff
@@ -169,14 +168,58 @@ namespace: #f
      ((symbolic? obj)
       (symbolic-hash obj))
      (else
-      (fxand
-       (let ((sn (object->serial-number obj)))
-         (if (fixnum? sn)
-           sn
-           (fxarithmetic-shift-left
-            (##bignum.mdigit-ref sn 0)
-            10)))
-       (macro-max-fixnum32))))))
+      (fxand (__eq-hash obj)
+             (macro-max-fixnum32))))))
+
+(cond-expand
+  (gerbil-smp
+   (def (__eq-hash obj)
+     (declare (not interrupts-enabled))
+     (let again ()
+       (if (##fx= (##vector-cas! __eq-hash-lock 0 1 0) 0)
+         (begin
+           (__object->eq-hash obj)
+           (##vector-cas! __eq-hash-lock 0 0 1))
+         (begin
```

Comment:
why `begin` here ?

Reply:
uhm, yes, why? me has no good answer :)
I will remove it but not regen bootstrap since it is the same code anyway.

Reply:
and the top arm is buggy, it will return the result of the vector cas... :man_facepalming:

Reply:
fixed. Note no need to bootstrap regen, as it is a UP compilation.

Speaking of which, we will need to set the core count to 1 when building stage1 on an SMP system, unless the bootstrap is compiled for SMP too. But we'll cross the bridge when we get there.